include/ruby/fiber/scheduler.h

```diff
@@ -0,0 +1,44 @@
+#ifndef RUBY_FIBER_SCHEDULER_H                                 /*-*-C-*-vi:se ft=c:*/
+#define RUBY_FIBER_SCHEDULER_H
+/**
+ * @file
+ * @author     Ruby developers <ruby-core@ruby-lang.org>
+ * @copyright  This  file  is   a  part  of  the   programming  language  Ruby.
+ *             Permission  is hereby  granted,  to  either redistribute  and/or
+ *             modify this file, provided that  the conditions mentioned in the
+ *             file COPYING are met.  Consult the file for details.
+ * @brief      Internal header for Scheduler.
+ */
+#include "ruby/ruby.h"
+#include "ruby/intern.h"
+
+VALUE rb_fiber_scheduler_get();
+VALUE rb_fiber_scheduler_set(VALUE scheduler);
+
+VALUE rb_fiber_scheduler_current();
+VALUE rb_thread_fiber_scheduler_current(VALUE thread);
+
+VALUE rb_fiber_scheduler_timeout(struct timeval *timeout);
+
+VALUE rb_fiber_scheduler_close(VALUE scheduler);
+
+VALUE rb_fiber_scheduler_kernel_sleep(VALUE scheduler, VALUE duration);
+VALUE rb_fiber_scheduler_kernel_sleepv(VALUE scheduler, int argc, VALUE * argv);
+
+int rb_fiber_scheduler_supports_process_wait(VALUE scheduler);
+VALUE rb_fiber_scheduler_process_wait(VALUE scheduler, rb_pid_t pid, int flags);
+
+VALUE rb_fiber_scheduler_block(VALUE scheduler, VALUE blocker, VALUE timeout);
+VALUE rb_fiber_scheduler_unblock(VALUE scheduler, VALUE blocker, VALUE fiber);
+
+VALUE rb_fiber_scheduler_io_wait(VALUE scheduler, VALUE io, VALUE events, VALUE timeout);
+VALUE rb_fiber_scheduler_io_wait_readable(VALUE scheduler, VALUE io);
+VALUE rb_fiber_scheduler_io_wait_writable(VALUE scheduler, VALUE io);
+
+int rb_fiber_scheduler_supports_io_read(VALUE scheduler);
+VALUE rb_fiber_scheduler_io_read(VALUE scheduler, VALUE io, VALUE buffer, size_t offset, size_t length);
+
+int rb_fiber_scheduler_supports_io_write(VALUE scheduler);
+VALUE rb_fiber_scheduler_io_write(VALUE scheduler, VALUE io, VALUE buffer, size_t offset, size_t length);
```

Comment:
Do we really want to expose all these C function as **public** C APIs?

The use case for https://github.com/bruno-/ruby/pull/1/files seems to just be able to:
`rb_respond_to(scheduler, id_address_resolve);`
and
`rb_funcall(scheduler, id_address_resolve, 2, hostname, timeout);`.

Why not just do that directly instead of exposing internal functions?

The scheduler method names are documented, so they are stable, making them public C APIs will make it less flexible to refactor that code.

Reply:
hot wo use C-APIs for socket? (small example is enough to explain why it is needed)

Reply:
I actually don't have any strong opinion about it, but this will be a problem for C extensions which want to invoke scheduler hooks. I don't want them using `rb_funcall` because it's less flexible for us in the future.

Reply:
Could you expose just the ones useful for C extension?

I think that's:
```
VALUE rb_fiber_scheduler_current();

VALUE rb_fiber_scheduler_timeout(struct timeval *timeout); // if useful

VALUE rb_fiber_scheduler_block(VALUE scheduler, VALUE blocker, VALUE timeout);
VALUE rb_fiber_scheduler_unblock(VALUE scheduler, VALUE blocker, VALUE fiber);

// IO
VALUE rb_fiber_scheduler_io_wait(VALUE scheduler, VALUE io, VALUE events, VALUE timeout);
VALUE rb_fiber_scheduler_io_wait_readable(VALUE scheduler, VALUE io);
VALUE rb_fiber_scheduler_io_wait_writable(VALUE scheduler, VALUE io);

int rb_fiber_scheduler_supports_io_read(VALUE scheduler);
VALUE rb_fiber_scheduler_io_read(VALUE scheduler, VALUE io, VALUE buffer, size_t offset, size_t length);

int rb_fiber_scheduler_supports_io_write(VALUE scheduler);
VALUE rb_fiber_scheduler_io_write(VALUE scheduler, VALUE io, VALUE buffer, size_t offset, size_t length);
```

And these I think should remain internal:
```
VALUE rb_fiber_scheduler_set(VALUE scheduler);

VALUE rb_fiber_scheduler_get();
VALUE rb_thread_fiber_scheduler_current(VALUE thread);

VALUE rb_fiber_scheduler_close(VALUE scheduler);

VALUE rb_fiber_scheduler_kernel_sleep(VALUE scheduler, VALUE duration);
VALUE rb_fiber_scheduler_kernel_sleepv(VALUE scheduler, int argc, VALUE * argv);

int rb_fiber_scheduler_supports_process_wait(VALUE scheduler);
VALUE rb_fiber_scheduler_process_wait(VALUE scheduler, rb_pid_t pid, int flags);
```

Reply:
This split seems fairly arbitrary to me. For example, why not sleep? But in any case, we generally don't want people calling these hooks unless they are augmenting an existing interface.

So, I'd be okay with having hooks which are only needed by extensions as public.

We generally don't want people "calling" hooks - they should be invoking the code paths in Ruby that already include this logic.

So, from that point of view, I'd be fine for this to be more limited - specifically to the functions needed by extensions (e.g. network name resolution for the socket extension).

The other option is for those extensions to define their own hooks. However, I think there will only ever be a finite number of hooks so we should probably enumerate them all in one place if possible.

Reply:
Sorry, there is actually one place where it makes sense to use the scheduler hooks directly.

If you are writing high performance code, and you want to invoke `rb_fiber_scheduler_io_wait_readable` multiple times in a tight loop, you can get the scheduler once, then invoke it multiple times to amortise the lookup cost. IMHO, this should be pretty small... to the point where it probably doesn't matter much. But at least it's one valid use case.

Reply:
> So, from that point of view, I'd be fine for this to be more limited - specifically to the functions needed by extensions (e.g. network name resolution for the socket extension).

Sounds good to me.

> rb_fiber_scheduler_io_wait_readable multiple times in a tight loop

Is there any change of calling rb_fiber_scheduler_io_wait_readable vs rb_funcall? I would think not.

Reply:
@ioquatix It seems you kept all functions as public? I see no reason to do so.

Reply:
> I don't want them using `rb_funcall` because it's less flexible for us in the future.

Could you expand a bit on why it's less flexible?

> How does it impact TruffleRuby?

If such new C APIs are introduced, it would be great to test them, if possible in ruby/spec C-API specs, which are by far the easiest to run tests for C APIs.
What I've seen so far is many C APIs are untested, and it's a pain to check compatibility since there is no proper test for them.